### ISSUE TITLE
"Magical Girl Pretty Sammy" by Wedge (1994)

### ISSUE THREAD

**Issue by emhoracek** [emhoracek/smooch#14]
Like #10, this doll is missing some items on certain set numbers. 

Set #5 in Smooch: 
<img width="659" alt="screen shot 2017-08-15 at 9 56 41 pm" src="https://user-images.githubusercontent.com/5353499/29344285-aa1f05b2-8204-11e7-8cb9-1890d8454976.png">

And in UltraKiss:
<img width="664" alt="screen shot 2017-08-15 at 8 51 00 pm" src="https://user-images.githubusercontent.com/5353499/29344289-b33d16fc-8204-11e7-9473-292b3cc4a33a.png">

On OtakuWorld: http://otakuworld.com/kiss/dolls/pages/s/sasami.htm

**Comment by emhoracek** [comment]
Looking at the [set data JSON](https://gist.github.com/emhoracek/869f78275d3479115f47b1854d52fd3f), it seems like a lot of cels don't have any position data associated with them. 

It would be great if I could find a much smaller set that had this same problem.

**Comment by huggablemonad** [comment]
This might be a red herring, but there's some kind of pattern in the errors that I'm seeing in the JavaScript console.

When I click on set 5, there's an `Uncaught TypeError`:

```
Uncaught TypeError: Cannot read property 'x' of undefined
    at KiSSCell.draw (doll.js:289)
    at KissSet.draw (doll.js:173)
    at HTMLAnchorElement.<anonymous> (doll.js:146)
```

Apparently, the position of \#64 (`sa74`) is undefined.

After removing all references to \#64 in `sasami.cnf`:

![sasami-set-5-no-64](https://user-images.githubusercontent.com/17948023/29517976-227ea518-86aa-11e7-98d6-ca89d563d005.png)

But I get the same error, only now it's \#63 (`sa71`) that has an undefined position. After removing all references to \#63 in `sasami.cnf`:

![sasami-set-5-no-64-63](https://user-images.githubusercontent.com/17948023/29518234-1ffbe9c6-86ab-11e7-89a7-6f4391563c3a.png)

Set 5 no longer gives me an error, but set 7 does. It's \#62 (`sa69`) with an undefined position. After removing all references to \#62 in `sasami.cnf`:

![sasami-set-7-no-62](https://user-images.githubusercontent.com/17948023/29518450-eeb920b2-86ab-11e7-9aa6-c09ea7ffdddd.png)

Set 7 still gives the same error, with \#61 (`sa68`) having an undefined position. I stopped debugging at this point.

I don't know if this helps any, but I thought I'd put it here anyway.

The diff with \#62 to \#64 removed:

```diff
--- sasami.cnf	2017-08-21 20:20:53.204280054 +0800
+++ sasami-no-64-63-62.cnf	2017-08-21 20:20:53.204280054 +0800
@@ -18,11 +18,9 @@
 ;
 #48	sa51.cel	:0 1 2 3 4 5 6 7 8	;タイトルプレート
 #60	sa67.cel	:              7 8	;首リボン
-#62	sa69.cel	:              7 8	;左手袋
 #61	sa68.cel	:              7 8	;右手袋
 #29	sa33.cel        :    2         7        ;左腕輪
 #27	sa31.cel        :    2         7        ;右腕輪
-#63.256	sa71.cel	:0 1 2 3 4 5 6 7 8	;前々髪
 #26	sa30.cel        :    2       6 7 8      ;うさ耳
 #13.256	sa13.cel	:0 1 2 3 4 5 6 7 8	;前髪
 #57	sa63.cel	:          5   7 	;リボン・右
@@ -62,7 +60,6 @@
 #49	sa52.cel	:0	   5 6 7	;ステッキ・上
 #3.256	sa03.cel	:0 1 2 3 4 5 6 7 8	;右腕
 #49	sa53.cel	:0	   5 6 7	;ステッキ・下
-#64	sa75.cel	:0   2     5   7	;サミー’Ｓスーツ・上
 #33	sa40.cel	:      3 4     7        ;スクール水着
 #32	sa39.cel	:      3     6 7	;水着
 #56	sa62.cel	:            6 7 8	;白タイツ
@@ -72,7 +69,6 @@
 #15	sa16.cel	:0 1 2 3 4 5 6 7 8	;パンティ
 #1.512	sa77.cel	:0 1 2 3 4 5 6 7 8	;砂沙美ちゃん・上
 ;#2.512	sa73.cel	:0 1 2 3 4 5 6 7 8	;秘密の砂沙美ちゃん・上
-#64	sa74.cel	:0  2      5   7 	;サミー’Ｓスーツ・下
 #1.512	sa76.cel	:0 1 2 3 4 5 6 7 8	;砂沙美ちゃん・下
 ;#2.512	sa72.cel	:0 1 2 3 4 5 6 7 8	;秘密の砂沙美ちゃん・下
 #60	sa70.cel	:              7 8	;首リボン（下）
@@ -94,39 +90,38 @@
 $0 * 90,28 * 118,94 * 189,72 137,71 190,271 132,303 211,305 126,355 128,15
  180,15 135,28 151,110 151,154 * * * * 353,48 383,51 * * * * * * * * *
  * * * * * * * * * * * 480,222 508,220 494,100 469,72 * * 395,297 76,13
- 118,94 329,86 342,140 380,225 * 337,235 * * * * * * * 137,30 90,21
+ 118,94 329,86 342,140 380,225 * 337,235 * * * * * *
 $0 * 90,28 * 118,94 * 189,74 138,69 * * * * * * 135,28 151,110 151,154
  166,244 132,259 121,90 210,305 132,22 182,24 * 130,361 * * * * * * * *
  * * * * * * * * 479,100 487,159 326,218 362,208 * * 515,227 475,231 395,297
- * 118,94 313,97 326,152 313,320 * 277,335 * * * * * * * 137,30
+ * 118,94 313,97 326,152 313,320 * 277,335 * * * * *
 $0 * 90,28 * 118,94 * 189,73 138,69 * * * * 317,44 386,47 135,28 151,110
  151,154 * * * * 132,22 182,24 * * 166,92 132,138 116,6 119,119 151,106
  227,158 127,365 211,301 * * * * * * * * 456,103 463,162 * * * * 490,227
- 449,239 395,297 * 118,94 * * * * * * * * * * * * 137,30 292,21
+ 449,239 395,297 * 118,94 * * * * * * * * * * *
 $0 * 90,28 * 118,94 * 189,74 138,69 * * * * * * 135,28 151,110 151,154
  * * * * 132,22 182,24 * * * * * * * * * * 448,89 151,93 * * * * * * *
  * * * 353,173 320,172 * * 395,297 * 118,94 307,66 321,121 497,218 * 449,218
- * * * * * * * 137,30
+ * * * * *
 $0 * 90,28 * 118,94 * 189,73 138,69 * * * * * * 135,28 150,110 151,154
  * * * * 132,22 182,24 * * * 511,115 * * * * * * * 445,97 * * * * * * 139,93
  145,152 303,261 338,230 166,244 132,259 210,305 131,361 395,297 * 118,94
- 298,90 312,146 353,255 * 301,277 * * * * * * * 137,30
+ 298,90 312,146 353,255 * 301,277 * * * * *
 $0 * 90,28 * 118,94 * 189,73 138,69 382,214 336,213 511,235 467,226 473,91
  507,94 135,28 151,110 151,154 * * * * 132,22 182,24 * * * * * * * * *
  * * * * * * * * * * * 132,333 196,280 * * * * 395,297 282,67 118,94 119,92
- 132,147 206,298 * 130,348 * 485,145 521,145 * * * * 137,30 305,11
+ 132,147 206,298 * 130,348 * 485,145 521,145 * * *
 $0 * 90,28 * 118,94 * 189,74 138,69 * * * * * * 135,28 151,110 151,154
  * * 303,73 491,218 132,22 182,24 * 446,223 * * 116,6 * 151,107 * * * 431,82
  * * * * * * * * * * * 336,176 314,180 * * 395,297 76,13 118,94 * * * *
- * 132,186 * * * * * * 137,30
+ * 132,186 * * * *
 $0 * 90,28 * 118,94 * 189,74 138,69 313,317 294,307 49,105 1,94 6,47 44,46
  135,28 150,110 151,154 286,198 282,174 377,62 58,343 132,22 182,24 * 8,356
  558,119 527,137 512,60 517,121 331,169 591,120 0,166 48,165 367,165 362,161
  * * * * * * 338,59 345,118 264,330 283,322 272,208 266,181 56,224 4,224
  395,297 76,13 118,94 266,56 278,112 51,274 * 3,279 132,186 220,213 246,214
- 463,75 475,50 510,141 586,141 137,30 382,86
+ 463,75 475,50 510,141
 $0 * 90,28 * 118,94 * 189,73 138,69 * * * * 299,52 342,52 135,28 151,110
  151,154 * * * * 132,22 183,24 * * * * 116,5 * * * * * * * * * * * * *
  * * * * * * * * 395,297 * 118,94 * * * * * 132,186 * * 143,128 157,88
- 119,105 205,135 137,30
 $0 *
```

**Comment by emhoracek** [comment]
Yeah I thought it probably had to do with that console error too. #10 has the same error and probably the same bug. Thanks for the information -- really interesting that if you remove objects the set starts to work better...

**Comment by huggablemonad** [comment]
I think I figured it out. If the length of a set is less than the length of the longest set, then the "missing" objects need to have a default position of `0,0` (cf. #15). In other words, for this doll, the longest set has 65 objects. So, for set 9, which is just `$0 *`, Smooch needs to treat it as being equivalent to `$0 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *` (65 asterisks).

The modified sets which I used for testing:
```
$0 * 90,28 * 118,94 * 189,72 137,71 190,271 132,303 211,305 126,355 128,15
 180,15 135,28 151,110 151,154 * * * * 353,48 383,51 * * * * * * * * *
 * * * * * * * * * * * 480,222 508,220 494,100 469,72 * * 395,297 76,13
 118,94 329,86 342,140 380,225 * 337,235 * * * * * * * 137,30 90,21
$0 * 90,28 * 118,94 * 189,74 138,69 * * * * * * 135,28 151,110 151,154
 166,244 132,259 121,90 210,305 132,22 182,24 * 130,361 * * * * * * * *
 * * * * * * * * 479,100 487,159 326,218 362,208 * * 515,227 475,231 395,297
 * 118,94 313,97 326,152 313,320 * 277,335 * * * * * * * 137,30 *
$0 * 90,28 * 118,94 * 189,73 138,69 * * * * 317,44 386,47 135,28 151,110
 151,154 * * * * 132,22 182,24 * * 166,92 132,138 116,6 119,119 151,106
 227,158 127,365 211,301 * * * * * * * * 456,103 463,162 * * * * 490,227
 449,239 395,297 * 118,94 * * * * * * * * * * * * 137,30 292,21
$0 * 90,28 * 118,94 * 189,74 138,69 * * * * * * 135,28 151,110 151,154
 * * * * 132,22 182,24 * * * * * * * * * * 448,89 151,93 * * * * * * *
 * * * 353,173 320,172 * * 395,297 * 118,94 307,66 321,121 497,218 * 449,218
 * * * * * * * 137,30 *
$0 * 90,28 * 118,94 * 189,73 138,69 * * * * * * 135,28 150,110 151,154
 * * * * 132,22 182,24 * * * 511,115 * * * * * * * 445,97 * * * * * * 139,93
 145,152 303,261 338,230 166,244 132,259 210,305 131,361 395,297 * 118,94
 298,90 312,146 353,255 * 301,277 * * * * * * * 137,30 *
$0 * 90,28 * 118,94 * 189,73 138,69 382,214 336,213 511,235 467,226 473,91
 507,94 135,28 151,110 151,154 * * * * 132,22 182,24 * * * * * * * * *
 * * * * * * * * * * * 132,333 196,280 * * * * 395,297 282,67 118,94 119,92
 132,147 206,298 * 130,348 * 485,145 521,145 * * * * 137,30 305,11
$0 * 90,28 * 118,94 * 189,74 138,69 * * * * * * 135,28 151,110 151,154
 * * 303,73 491,218 132,22 182,24 * 446,223 * * 116,6 * 151,107 * * * 431,82
 * * * * * * * * * * * 336,176 314,180 * * 395,297 76,13 118,94 * * * *
 * 132,186 * * * * * * 137,30 *
$0 * 90,28 * 118,94 * 189,74 138,69 313,317 294,307 49,105 1,94 6,47 44,46
 135,28 150,110 151,154 286,198 282,174 377,62 58,343 132,22 182,24 * 8,356
 558,119 527,137 512,60 517,121 331,169 591,120 0,166 48,165 367,165 362,161
 * * * * * * 338,59 345,118 264,330 283,322 272,208 266,181 56,224 4,224
 395,297 76,13 118,94 266,56 278,112 51,274 * 3,279 132,186 220,213 246,214
 463,75 475,50 510,141 586,141 137,30 382,86
$0 * 90,28 * 118,94 * 189,73 138,69 * * * * 299,52 342,52 135,28 151,110
 151,154 * * * * 132,22 183,24 * * * * 116,5 * * * * * * * * * * * * *
 * * * * * * * * 395,297 * 118,94 * * * * * 132,186 * * 143,128 157,88
 119,105 205,135 137,30 *
$0 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
```

**Comment by huggablemonad** [comment]
A minimal reproducible example:

```
; sasami.cnf
;
; A minimal reproducible example for issue #14, where there're items missing in
; certain sets.

(640,400)
%sasami.kcf
[9

#2     sa51.cel :0 1
#1.512 sa77.cel :0 1

$0 * 90,28

$0 * 90,28 395,297
```

Set 0 in GnomeKiSS:
![gnomekiss-set-0](https://user-images.githubusercontent.com/17948023/44625606-e97b3600-a93f-11e8-8086-01a2ca775c77.jpg)

Set 0 in Smooch:
![smooch-set-0](https://user-images.githubusercontent.com/17948023/44625618-3b23c080-a940-11e8-95e0-4c3c7f906e2f.jpg)

Set 1 in GnomeKiSS:
![gnomekiss-set-1](https://user-images.githubusercontent.com/17948023/44625607-f009ad80-a93f-11e8-9699-c88b19e1d6fe.jpg)

Set 1 in Smooch:
![smooch-set-1](https://user-images.githubusercontent.com/17948023/44625623-42e36500-a940-11e8-9d6a-bca48d9d12c4.jpg)

**Comment by emhoracek** [comment]
This was also fixed by #93 !
